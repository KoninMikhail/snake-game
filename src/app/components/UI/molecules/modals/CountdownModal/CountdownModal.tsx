import {useCountdown} from 'usehooks-ts';
import {
    StyledCountdownModal,
    StyledCountdownModalContent,
} from '@ui/molecules/modals/CountdownModal/style/StyledCountdownModal';
import {useEffect} from 'react';
import {HeavyCaption} from '@ui/atoms/typography/caption/HeavyCaption/HeavyCaption';
import {Text} from '@ui/atoms/typography/body/Text/Text';
import {ICountdownModal} from '@ui/molecules/modals/CountdownModal/CountdownModalProps.interface';

/**
 * @name CountdownModal
 * @author Mikhail Konin <dev.konin@gmail.com>
 *
 * Use to ready a message about incorrect screen orientation.
 *
 * @param isShow - showing state
 * @param countStart - count from number
 * @param onCountdownEnd - callback
 * @returns {JSX.Element}
 */
export const CountdownModal = <T extends number>({
    visible,
    message,
    countStart,
    onCountdownEnd,
}: ICountdownModal<T>) => {
    const [count, { startCountdown, resetCountdown }] = useCountdown({
        countStart,
        intervalMs: 1000,
    });

    useEffect(() => {
        if (visible) {
            startCountdown();
        }
        if (count === 0) {
            onCountdownEnd && onCountdownEnd();
            resetCountdown();
        }
    }, [count, onCountdownEnd, resetCountdown, startCountdown, visible]);

    return (
        <>
            <StyledCountdownModal>
                <StyledCountdownModalContent>
                    <HeavyCaption align={'center'}>{count}</HeavyCaption>
                    <Text align={'center'}>{message}</Text>
                </StyledCountdownModalContent>
            </StyledCountdownModal>
        </>
    );
};
