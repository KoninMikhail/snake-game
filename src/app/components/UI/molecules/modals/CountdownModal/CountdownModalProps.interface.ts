import { PositiveInteger } from '@/types/positive-integer';

export interface ICountdownModal<T extends PositiveInteger<number>> {
    visible: boolean;
    message: string;
    countStart: T;
    onCountdownEnd?: () => void;
}
