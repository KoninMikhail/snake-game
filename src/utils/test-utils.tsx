import React, {PropsWithChildren} from 'react';
import type {RenderOptions} from '@testing-library/react';
import {render} from '@testing-library/react';
import type {PreloadedState} from '@reduxjs/toolkit';
import {configureStore} from '@reduxjs/toolkit';
import {Provider} from 'react-redux';

import type {AppStore, RootState} from '@store/store';
import appSlice from '@store/slices/app.slice';
import soundSlice from '@store/slices/sound.slice';
import maxScoreSlice from '@store/slices/maxScore.slice';
import gameDataSlice from '@store/slices/game.slice';
import {ImagesData} from '@data/imagesData';
import {mockImagesData, mockSoundData} from '@/setupTests';
import SoundsProvider from '@context/SoundsContext';
import ImagesProvider from '@context/ImagesContext';
import {SoundsData} from '@data/soundsData';
import {DefaultTheme, ThemeProvider} from 'styled-components';
import {theme as appTheme} from '@styles/theme';

// As a basic setup, import your same slice reducers

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
    preloadedState?: PreloadedState<RootState>;
    store?: AppStore;
    images?: ImagesData;
    sounds?: SoundsData;
    theme?: DefaultTheme;
}

export function renderWithProviders(
    ui: React.ReactElement,
    {
        preloadedState = {},
        // Automatically create a store instance if no store was passed in
        store = configureStore({
            reducer: {
                app: appSlice,
                sounds: soundSlice,
                maxScore: maxScoreSlice,
                game: gameDataSlice,
            },
            preloadedState,
        }),
        images = mockImagesData,
        sounds = mockSoundData,
        theme = appTheme,
        ...renderOptions
    }: ExtendedRenderOptions = {}
) {
    function Wrapper({children}: PropsWithChildren): JSX.Element {
        return (
            <Provider store={store}>
                <ThemeProvider theme={theme}>
                    <SoundsProvider sounds={sounds}>
                        <ImagesProvider images={images}>{children}</ImagesProvider>
                    </SoundsProvider>
                </ThemeProvider>
            </Provider>
        );
    }

    // Return an object with the store and all of RTL's query functions
    return {store, ...render(ui, {wrapper: Wrapper, ...renderOptions})};
}
