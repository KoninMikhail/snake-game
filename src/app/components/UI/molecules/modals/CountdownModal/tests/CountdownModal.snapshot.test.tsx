import { CountdownModal } from '@ui/molecules/modals/CountdownModal/CountdownModal';
import { renderWithProviders } from '@utils/test-utils';

/**
 * Test
 */
describe('CountdownModal', () => {
    it('equal snapshot when showed', async () => {
        const { container } = renderWithProviders(
            <CountdownModal visible={true} countStart={5} message={'message'} />
        );

        expect(container).toMatchSnapshot();
    });

    it('equal snapshot when hidden', async () => {
        const { container } = renderWithProviders(
            <CountdownModal visible={false} countStart={1} message={'message'} />
        );

        expect(container).toMatchSnapshot();
    });
});
