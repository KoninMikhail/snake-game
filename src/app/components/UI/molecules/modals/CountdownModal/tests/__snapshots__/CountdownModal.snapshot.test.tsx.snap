// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountdownModal equal snapshot when hidden 1`] = `
.c0 {
  position: absolute;
  display: grid;
  top: 0;
  left: 0;
  bottom: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 15px;
  width: 100%;
  z-index: 9010;
  height: 100%;
  background: black;
}

.c1 {
  display: grid;
  gap: 1.5rem;
}

.c2 {
  font-family: 'pixeloid_bold','Segoe UI','Roboto','Oxygen','Ubuntu','Fira Sans','Droid Sans','Helvetica Neue',sans-serif;
  font-size: 2.8em;
  font-weight: normal;
  text-align: center;
}

.c3 {
  color: white;
  font-size: 14px;
  font-family: 'pixeloid_bold','Segoe UI','Roboto','Oxygen','Ubuntu','Fira Sans','Droid Sans','Helvetica Neue',sans-serif;
  font-weight: normal;
  text-align: center;
}

@media (min-width:768px) {
  .c2 {
    font-size: 2.5em;
  }
}

@media (min-width:1200px) {
  .c2 {
    font-size: 3.4em;
  }
}

@media (min-width:768px) {
  .c3 {
    font-size: 16px;
  }
}

@media (min-width:1200px) {
  .c3 {
    font-size: 1.1vw;
  }
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <span
        class="c2"
      >
        1
      </span>
      <span
        class="c3"
      >
        message
      </span>
    </div>
  </div>
</div>
`;

exports[`CountdownModal equal snapshot when showed 1`] = `
.c0 {
  position: absolute;
  display: grid;
  top: 0;
  left: 0;
  bottom: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 15px;
  width: 100%;
  z-index: 9010;
  height: 100%;
  background: black;
}

.c1 {
  display: grid;
  gap: 1.5rem;
}

.c2 {
  font-family: 'pixeloid_bold','Segoe UI','Roboto','Oxygen','Ubuntu','Fira Sans','Droid Sans','Helvetica Neue',sans-serif;
  font-size: 2.8em;
  font-weight: normal;
  text-align: center;
}

.c3 {
  color: white;
  font-size: 14px;
  font-family: 'pixeloid_bold','Segoe UI','Roboto','Oxygen','Ubuntu','Fira Sans','Droid Sans','Helvetica Neue',sans-serif;
  font-weight: normal;
  text-align: center;
}

@media (min-width:768px) {
  .c2 {
    font-size: 2.5em;
  }
}

@media (min-width:1200px) {
  .c2 {
    font-size: 3.4em;
  }
}

@media (min-width:768px) {
  .c3 {
    font-size: 16px;
  }
}

@media (min-width:1200px) {
  .c3 {
    font-size: 1.1vw;
  }
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <span
        class="c2"
      >
        5
      </span>
      <span
        class="c3"
      >
        message
      </span>
    </div>
  </div>
</div>
`;
